File: workers/automation/src/jobhunter/domain/events/discovery.py
"""Job Discovery domain events.

See ddd-target.md §4.1.
"""

from __future__ import annotations

from dataclasses import asdict, dataclass, field
from typing import Any

from jobhunter.domain.tenant import TenantId
from jobhunter.domain.events.base import DomainEvent, create_domain_event


@dataclass(frozen=True)
class JobDiscoveredPayload:
    job_id: str
    posting_url: str
    source: str
    employer: str
    metadata: dict[str, Any] = field(default_factory=dict)
    discovered_at: str = ""


def create_job_discovered(tenant_id: TenantId, payload: JobDiscoveredPayload) -> DomainEvent:
    return create_domain_event("JobDiscovered", tenant_id, asdict(payload))


@dataclass(frozen=True)
class JobUpdatedPayload:
    job_id: str
    changed_fields: dict[str, Any] = field(default_factory=dict)


def create_job_updated(tenant_id: TenantId, payload: JobUpdatedPayload) -> DomainEvent:
    return create_domain_event("JobUpdated", tenant_id, asdict(payload))


@dataclass(frozen=True)
class JobDeletedPayload:
    job_id: str
    reason: str
    deleted_at: str


def create_job_deleted(tenant_id: TenantId, payload: JobDeletedPayload) -> DomainEvent:
    return create_domain_event("JobDeleted", tenant_id, asdict(payload))


@dataclass(frozen=True)
class JobRestoredPayload:
    job_id: str
    restored_at: str


def create_job_restored(tenant_id: TenantId, payload: JobRestoredPayload) -> DomainEvent:
    return create_domain_event("JobRestored", tenant_id, asdict(payload))


@dataclass(frozen=True)
class SourceLocationCandidateDiscoveredPayload:
    candidate_id: str
    candidate_url: str
    source_kind: str
    confidence: float
    evidence_ref: str
    discovered_at: str


def create_source_location_candidate_discovered(
    tenant_id: TenantId,
    payload: SourceLocationCandidateDiscoveredPayload,
) -> DomainEvent:
    return create_domain_event("SourceLocationCandidateDiscovered", tenant_id, asdict(payload))


@dataclass(frozen=True)
class SourceLocationCandidatePromotedPayload:
    candidate_id: str
    source_id: str
    promoted_at: str


def create_source_location_candidate_promoted(
    tenant_id: TenantId,
    payload: SourceLocationCandidatePromotedPayload,
) -> DomainEvent:
    return create_domain_event("SourceLocationCandidatePromoted", tenant_id, asdict(payload))


@dataclass(frozen=True)
class SourceRegistryEntryCreatedPayload:
    source_id: str
    kind: str
    policy_id: str
    state: str
    created_at: str


def create_source_registry_entry_created(
    tenant_id: TenantId,
    payload: SourceRegistryEntryCreatedPayload,
) -> DomainEvent:
    return create_domain_event("SourceRegistryEntryCreated", tenant_id, asdict(payload))


@dataclass(frozen=True)
class SourceRegistryEntryUpdatedPayload:
    source_id: str
    changed_fields: tuple[str, ...]
    updated_at: str


def create_source_registry_entry_updated(
    tenant_id: TenantId,
    payload: SourceRegistryEntryUpdatedPayload,
) -> DomainEvent:
    return create_domain_event("SourceRegistryEntryUpdated", tenant_id, asdict(payload))


@dataclass(frozen=True)
class SourceStateChangedPayload:
    source_id: str
    from_state: str
    to_state: str
    reason: str
    changed_at: str


def create_source_state_changed(
    tenant_id: TenantId,
    payload: SourceStateChangedPayload,
) -> DomainEvent:
    return create_domain_event("SourceStateChanged", tenant_id, asdict(payload))

File: packages/domain-types/src/events/discovery.ts
/**
 * Job Discovery domain events.
 *
 * @see ddd-target.md §4.1
 */

import type { TenantId } from "../tenant.js";
import type { SourceKind, SourceState } from "../discovery/source.js";
import { type DomainEvent, createDomainEvent } from "./base.js";

// -- JobDiscovered ----------------------------------------------------------

export interface JobDiscoveredPayload {
  readonly jobId: string;
  readonly postingUrl: string;
  readonly source: string;
  readonly employer: string;
  readonly metadata: Record<string, unknown>;
  readonly discoveredAt: string;
}

export type JobDiscovered = DomainEvent<"JobDiscovered", JobDiscoveredPayload>;

export function createJobDiscovered(
  tenantId: TenantId,
  payload: JobDiscoveredPayload,
): JobDiscovered {
  return createDomainEvent("JobDiscovered", tenantId, payload);
}

// -- JobUpdated -------------------------------------------------------------

export interface JobUpdatedPayload {
  readonly jobId: string;
  readonly changedFields: Record<string, unknown>;
}

export type JobUpdated = DomainEvent<"JobUpdated", JobUpdatedPayload>;

export function createJobUpdated(
  tenantId: TenantId,
  payload: JobUpdatedPayload,
): JobUpdated {
  return createDomainEvent("JobUpdated", tenantId, payload);
}

// -- JobDeleted -------------------------------------------------------------

export interface JobDeletedPayload {
  readonly jobId: string;
  readonly reason: string;
  readonly deletedAt: string;
}

export type JobDeleted = DomainEvent<"JobDeleted", JobDeletedPayload>;

export function createJobDeleted(
  tenantId: TenantId,
  payload: JobDeletedPayload,
): JobDeleted {
  return createDomainEvent("JobDeleted", tenantId, payload);
}

// -- JobRestored ------------------------------------------------------------

export interface JobRestoredPayload {
  readonly jobId: string;
  readonly restoredAt: string;
}

export type JobRestored = DomainEvent<"JobRestored", JobRestoredPayload>;

export function createJobRestored(
  tenantId: TenantId,
  payload: JobRestoredPayload,
): JobRestored {
  return createDomainEvent("JobRestored", tenantId, payload);
}

// -- SourceLocationCandidateDiscovered -------------------------------------

export interface SourceLocationCandidateDiscoveredPayload {
  readonly candidateId: string;
  readonly candidateUrl: string;
  readonly sourceKind: SourceKind;
  readonly confidence: number;
  readonly evidenceRef: string;
  readonly discoveredAt: string;
}

export type SourceLocationCandidateDiscovered = DomainEvent<
  "SourceLocationCandidateDiscovered",
  SourceLocationCandidateDiscoveredPayload
>;

export function createSourceLocationCandidateDiscovered(
  tenantId: TenantId,
  payload: SourceLocationCandidateDiscoveredPayload,
): SourceLocationCandidateDiscovered {
  return createDomainEvent("SourceLocationCandidateDiscovered", tenantId, payload);
}

// -- SourceLocationCandidatePromoted ---------------------------------------

export interface SourceLocationCandidatePromotedPayload {
  readonly candidateId: string;
  readonly sourceId: string;
  readonly promotedAt: string;
}

export type SourceLocationCandidatePromoted = DomainEvent<
  "SourceLocationCandidatePromoted",
  SourceLocationCandidatePromotedPayload
>;

export function createSourceLocationCandidatePromoted(
  tenantId: TenantId,
  payload: SourceLocationCandidatePromotedPayload,
): SourceLocationCandidatePromoted {
  return createDomainEvent("SourceLocationCandidatePromoted", tenantId, payload);
}

// -- SourceRegistryEntryCreated --------------------------------------------

export interface SourceRegistryEntryCreatedPayload {
  readonly sourceId: string;
  readonly kind: SourceKind;
  readonly policyId: string;
  readonly state: SourceState;
  readonly createdAt: string;
}

export type SourceRegistryEntryCreated = DomainEvent<
  "SourceRegistryEntryCreated",
  SourceRegistryEntryCreatedPayload
>;

export function createSourceRegistryEntryCreated(
  tenantId: TenantId,
  payload: SourceRegistryEntryCreatedPayload,
): SourceRegistryEntryCreated {
  return createDomainEvent("SourceRegistryEntryCreated", tenantId, payload);
}

// -- SourceRegistryEntryUpdated --------------------------------------------

export interface SourceRegistryEntryUpdatedPayload {
  readonly sourceId: string;
  readonly changedFields: readonly string[];
  readonly updatedAt: string;
}

export type SourceRegistryEntryUpdated = DomainEvent<
  "SourceRegistryEntryUpdated",
  SourceRegistryEntryUpdatedPayload
>;

export function createSourceRegistryEntryUpdated(
  tenantId: TenantId,
  payload: SourceRegistryEntryUpdatedPayload,
): SourceRegistryEntryUpdated {
  return createDomainEvent("SourceRegistryEntryUpdated", tenantId, payload);
}

// -- SourceStateChanged -----------------------------------------------------

export interface SourceStateChangedPayload {
  readonly sourceId: string;
  readonly fromState: SourceState;
  readonly toState: SourceState;
  readonly reason: string;
  readonly changedAt: string;
}

export type SourceStateChanged = DomainEvent<"SourceStateChanged", SourceStateChangedPayload>;

export function createSourceStateChanged(
  tenantId: TenantId,
  payload: SourceStateChangedPayload,
): SourceStateChanged {
  return createDomainEvent("SourceStateChanged", tenantId, payload);
}
